File: 第二阶段/2.5-装饰器/Decorators.cs
// ============================================================================
// C# Attribute 与拦截器 —— 对应 Python 装饰器
// 对应文章：2.5 装饰器
// ============================================================================
//
// 【C# 程序员的 Python 修炼手册】
//
// 这个文件展示 C# 中与 Python 装饰器等价的各种模式：
//   1. Attribute（元数据标注）      ← 最基础，但本身不能拦截调用
//   2. 装饰器模式（Decorator Pattern）← GoF 经典设计模式
//   3. 拦截器（Method Interception） ← Castle DynamicProxy / RealProxy
//   4. Action Filter               ← ASP.NET Core 最接近 Python 装饰器
//
// 核心区别：
// - Python 装饰器：@ 语法糖直接包装函数，运行时生效
// - C# Attribute：需要框架（ASP.NET Core、Castle）配合反射读取才能拦截
// - C# 更安全（类型检查），Python 更灵活（动态包装）
// ============================================================================

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Threading;

// ============================================================================
// 1. 基础 Attribute —— Python @decorator 的最简等价物
// ============================================================================
//
// 【Python 等价】
// def log(func):
//     def wrapper(*args, **kwargs):
//         print(f"调用 {func.__name__}")
//         return func(*args, **kwargs)
//     return wrapper
//
// 重要区别：
// - Python 装饰器在定义时就包装了函数（编译期等价）
// - C# Attribute 只是元数据标签，需要反射在运行时读取
// - 仅贴上 [Log] 标签不会产生任何效果！

/// <summary>
/// 日志 Attribute —— 标记需要记录日志的方法
/// 【对比 Python @timer / @log 装饰器】
/// </summary>
[AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
public class LogAttribute : Attribute
{
    public string Message { get; set; }

    public LogAttribute(string message = "")
    {
        Message = message;
    }
}

/// <summary>
/// 缓存 Attribute
/// 【对比 Python @cache(maxsize=128) 装饰器】
/// </summary>
[AttributeUsage(AttributeTargets.Method)]
public class CacheAttribute : Attribute
{
    public int DurationSeconds { get; set; } = 60;
}

/// <summary>
/// 权限检查 Attribute
/// 【对比 Python @require_role("Admin") 装饰器】
/// </summary>
[AttributeUsage(AttributeTargets.Method)]
public class AuthorizeAttribute : Attribute
{
    public string Roles { get; set; } = "";
}

/// <summary>
/// 计时 Attribute
/// 【对比 Python @timer 装饰器】
/// </summary>
[AttributeUsage(AttributeTargets.Method)]
public class TimedAttribute : Attribute { }


// ============================================================================
// 2. 方法拦截器（Method Interceptor）—— 真正实现"装饰"效果
// ============================================================================
//
// 【Python 等价】
// Python 装饰器本质上就是一个高阶函数包装：
//   @timer
//   def func(): ...
//   # 等价于 func = timer(func)
//
// C# 中要实现同样的效果，需要用接口 + 反射 + 代理
// 框架层面：Castle DynamicProxy, Autofac Interceptor, DispatchProxy
//
// 这里用 DispatchProxy（.NET 内置）演示核心原理

/// <summary>
/// 方法拦截信息 —— 对应 Python 装饰器 wrapper 函数的参数
/// </summary>
public class InterceptedMethod
{
    public string MethodName { get; set; } = "";
    public object?[] Args { get; set; } = Array.Empty<object?>();
    public Stopwatch Timer { get; set; } = new();
}

/// <summary>
/// 拦截代理 —— C# 版的"函数包装器"
/// 【核心概念】这等价于 Python 装饰器的 wrapper 函数
///
/// Python:
///   def wrapper(*args, **kwargs):
///       print(f"调用 {func.__name__}")
///       result = func(*args, **kwargs)
///       print(f"返回 {result}")
///       return result
///
/// C# (DispatchProxy):
///   protected override object? Invoke(MethodInfo targetMethod, object?[] args)
///   {
///       // 在目标方法前后添加逻辑 = 装饰器的 wrapper
///   }
/// </summary>
public class MethodInterceptionProxy<T> : DispatchProxy where T : class
{
    private T? _decorated = null!;
    private readonly List<Action<InterceptedMethod>> _interceptors = new();

    /// <summary>
    /// 注册拦截行为 —— 类似于给同一个函数堆叠多个 Python 装饰器
    /// </summary>
    public void AddInterceptor(Action<InterceptedMethod> interceptor)
    {
        _interceptors.Add(interceptor);
    }

    protected override object? Invoke(MethodInfo targetMethod, object?[]? args)
    {
        var info = new InterceptedMethod
        {
            MethodName = targetMethod.Name,
            Args = args ?? Array.Empty<object?>()
        };

        // 前置拦截（对应 Python 装饰器中 func 调用前的代码）
        foreach (var interceptor in _interceptors)
        {
            interceptor(info);
        }

        info.Timer.Start();
        try
        {
            // 调用原始方法
            var result = targetMethod.Invoke(_decorated, args);
            info.Timer.Stop();

            // 后置拦截（对应 Python 装饰器中 func 调用后的代码）
            Console.WriteLine($"  [C# Interceptor] {targetMethod.Name} 完成，" +
                            $"耗时 {info.Timer.ElapsedMilliseconds}ms，返回 {result}");
            return result;
        }
        catch (TargetInvocationException ex) when (ex.InnerException != null)
        {
            info.Timer.Stop();
            Console.WriteLine($"  [C# Interceptor] {targetMethod.Name} 异常: {ex.InnerException.Message}");
            throw ex.InnerException;
        }
    }

    /// <summary>
    /// 创建代理实例
    /// </summary>
    public static T Create(T decorated)
    {
        var proxy = Create<T, MethodInterceptionProxy<T>>() as MethodInterceptionProxy<T>;
        proxy!._decorated = decorated;
        return proxy as T ?? throw new InvalidOperationException("无法创建代理");
    }
}


// ============================================================================
// 3. Action Filter 模式 —— ASP.NET Core 风格
// ============================================================================
//
// 这是 ASP.NET Core 中最接近 Python 装饰器的模式
// 每个 Filter 就是一个装饰器，可以堆叠

/// <summary>
/// Filter 执行上下文
/// </summary>
public class FilterContext
{
    public string MethodName { get; set; } = "";
    public Dictionary<string, object?> Parameters { get; set; } = new();
    public bool IsCancelled { get; set; }
    public string? CancellationReason { get; set; }
}

/// <summary>
/// IActionFilter —— 等价于 Python 装饰器的接口
///
/// Python:
///   def decorator(func):
///       def wrapper(*args, **kwargs):
///           # Before
///           result = func(*args, **kwargs)
///           # After
///           return result
///       return wrapper
///
/// C#:
///   void OnBefore(FilterContext context);   // 函数调用前
///   void OnAfter(FilterContext context);     // 函数调用后
/// </summary>
public interface IActionFilter
{
    void OnBefore(FilterContext context);
    void OnAfter(FilterContext context);
}

/// <summary>
/// 计时 Filter —— 等价于 Python @timer
/// </summary>
public class TimerFilter : IActionFilter
{
    private Stopwatch? _stopwatch;

    public void OnBefore(FilterContext context)
    {
        _stopwatch = Stopwatch.StartNew();
        Console.WriteLine($"  [TimerFilter] 开始计时: {context.MethodName}");
    }

    public void OnAfter(FilterContext context)
    {
        _stopwatch?.Stop();
        Console.WriteLine($"  [TimerFilter] {context.MethodName} 耗时: {_stopwatch?.ElapsedMilliseconds}ms");
    }
}

/// <summary>
/// 日志 Filter —— 等价于 Python @log
/// </summary>
public class LogFilter : IActionFilter
{
    public void OnBefore(FilterContext context)
    {
        var paramStr = string.Join(", ", context.Parameters.Select(p => $"{p.Key}={p.Value}"));
        Console.WriteLine($"  [LogFilter] 调用 {context.MethodName}({paramStr})");
    }

    public void OnAfter(FilterContext context)
    {
        Console.WriteLine($"  [LogFilter] {context.MethodName} 执行完成");
    }
}

/// <summary>
/// 权限 Filter —— 等价于 Python @require_role
/// </summary>
public class AuthorizeFilter : IActionFilter
{
    private readonly string _requiredRole;

    public AuthorizeFilter(string requiredRole)
    {
        _requiredRole = requiredRole;
    }

    public void OnBefore(FilterContext context)
    {
        if (context.Parameters.TryGetValue("userRole", out var role))
        {
            if (role?.ToString() != _requiredRole)
            {
                context.IsCancelled = true;
                context.CancellationReason = $"需要 {_requiredRole} 角色，当前角色: {role}";
                Console.WriteLine($"  [AuthorizeFilter] 权限不足: {context.CancellationReason}");
            }
        }
    }

    public void OnAfter(FilterContext context) { }
}


// ============================================================================
// 4. Filter Pipeline 执行器 —— 等价于 Python 装饰器堆叠
// ============================================================================
//
// 【Python 堆叠装饰器】
//   @timer          ← 最外层，最后包装
//   @log            ← 中间层
//   @authorize      ← 最内层，最先包装
//   def func(): ...
//
// 调用时: timer.wrapper → log.wrapper → authorize.wrapper → func
//
// C# Filter Pipeline 是同样的思路，只是实现方式不同

/// <summary>
/// Filter Pipeline —— 管理多个 Action Filter 的执行
/// </summary>
public static class FilterPipeline
{
    public static T Execute<T>(Func<T> action, string methodName,
        Dictionary<string, object?>? parameters = null,
        params IActionFilter[] filters)
    {
        var context = new FilterContext
        {
            MethodName = methodName,
            Parameters = parameters ?? new()
        };

        // 执行所有 OnBefore（对应 Python 装饰器从外到内执行）
        foreach (var filter in filters)
        {
            filter.OnBefore(context);
            if (context.IsCancelled)
            {
                Console.WriteLine($"  [Pipeline] 请求被取消: {context.CancellationReason}");
                return default!;
            }
        }

        try
        {
            // 执行目标方法
            var result = action();

            // 执行所有 OnAfter（反向，对应 Python 装饰器返回时从内到外）
            foreach (var filter in filters.Reverse())
            {
                filter.OnAfter(context);
            }

            return result;
        }
        catch (Exception ex)
        {
            Console.WriteLine($"  [Pipeline] 方法执行异常: {ex.Message}");
            throw;
        }
    }
}


// ============================================================================
// 5. 装饰器模式（Decorator Pattern）—— GoF 设计模式
// ============================================================================
//
// 这是最传统的 C# 装饰方式，虽然啰嗦但类型安全

/// <summary>
/// 接口定义
/// </summary>
public interface IDataService
{
    string GetData(int id);
}

/// <summary>
/// 基础服务 —— 等价于 Python 的原始函数
/// </summary>
public class DataService : IDataService
{
    public string GetData(int id)
    {
        Console.WriteLine($"  [DataService] 从数据库获取 id={id}");
        return $"Data-{id}";
    }
}

/// <summary>
/// 缓存装饰器 —— 等价于 Python @cache
/// </summary>
public class CachedDataService : IDataService
{
    private readonly IDataService _inner;
    private readonly Dictionary<int, string> _cache = new();

    public CachedDataService(IDataService inner)
    {
        _inner = inner;  // 保存原始服务引用（类似 Python __wrapped__）
    }

    public string GetData(int id)
    {
        if (_cache.TryGetValue(id, out var cached))
        {
            Console.WriteLine($"  [Cached] id={id} 命中缓存");
            return cached;
        }

        var result = _inner.GetData(id);
        _cache[id] = result;
        return result;
    }
}

/// <summary>
/// 日志装饰器 —— 等价于 Python @timer / @log
/// </summary>
public class LoggingDataService : IDataService
{
    private readonly IDataService _inner;

    public LoggingDataService(IDataService inner)
    {
        _inner = inner;
    }

    public string GetData(int id)
    {
        var sw = Stopwatch.StartNew();
        Console.WriteLine($"  [Logging] GetData({id}) 开始");
        var result = _inner.GetData(id);
        sw.Stop();
        Console.WriteLine($"  [Logging] GetData({id}) 完成，耗时 {sw.ElapsedMilliseconds}ms");
        return result;
    }
}


// ============================================================================
// 6. 主程序入口
// ============================================================================

class Program
{
    static void Main(string[] args)
    {
        Console.WriteLine(new string('=', 70));
        Console.WriteLine("  C# Attribute 与拦截器 —— Python 装饰器的 C# 等价物");
        Console.WriteLine("  C# 程序员的 Python 修炼手册 —— 2.5");
        Console.WriteLine(new string('=', 70));

        // --- 1. Attribute 的本质：只是元数据标签 ---
        Console.WriteLine("\n--- 1. Attribute 本质：元数据标签（需要反射才能生效）---");
        Console.WriteLine("  [Log], [Cache], [Authorize] 只是标签，单独使用不会拦截任何调用");
        Console.WriteLine("  必须配合反射或框架（如 ASP.NET Core）才能产生实际效果\n");

        // 通过反射读取 Attribute —— 框架的工作
        var method = typeof(Program).GetMethod(nameof(DemoMethod))!;
        var attrs = method.GetCustomAttributes<LogAttribute>(inherit: true);
        Console.WriteLine($"  DemoMethod 上的 [Log] 属性数量: {attrs.Count()}");
        foreach (var attr in attrs)
        {
            Console.WriteLine($"    Message = \"{attr.Message}\"");
        }

        // --- 2. Decorator Pattern（经典装饰器模式）---
        Console.WriteLine("\n--- 2. Decorator Pattern（经典 GoF 装饰器）---");
        Console.WriteLine("  这是最传统的 C# 装饰方式：用接口 + 组合\n");

        // 链式包装：DataService → CachedDataService → LoggingDataService
        // 等价于 Python 的 @timer @cache def func()
        IDataService service = new DataService();
        service = new CachedDataService(service);  // 内层
        service = new LoggingDataService(service);  // 外层

        Console.WriteLine("  第一次调用（缓存未命中，穿透到数据库）:");
        Console.WriteLine($"  结果: {service.GetData(1)}");

        Console.WriteLine("\n  第二次调用（缓存命中）:");
        Console.WriteLine($"  结果: {service.GetData(1)}");

        // --- 3. Action Filter Pipeline（ASP.NET Core 风格）---
        Console.WriteLine("\n--- 3. Action Filter Pipeline（ASP.NET Core 风格）---");
        Console.WriteLine("  这是 ASP.NET Core 最接近 Python 装饰器的模式\n");

        // 场景1: 正常执行（timer + log 两个 filter 堆叠）
        Console.WriteLine("  场景1: 正常调用（timer + log filter）");
        var result = FilterPipeline.Execute(
            () =>
            {
                Thread.Sleep(10);  // 模拟耗时操作
                return 42;
            },
            "CalculateResult",
            new Dictionary<string, object?> { { "input", 10 } },
            new TimerFilter(),     // 装饰器1
            new LogFilter()        // 装饰器2
        );
        Console.WriteLine($"  返回值: {result}");

        // 场景2: 权限检查失败
        Console.WriteLine("\n  场景2: 权限检查（authorize + timer filter）");
        var deleteResult = FilterPipeline.Execute(
            () =>
            {
                Console.WriteLine("  [执行] 用户已被删除");
                return true;
            },
            "DeleteUser",
            new Dictionary<string, object?> { { "userId", 42 }, { "userRole", "Viewer" } },
            new AuthorizeFilter("Admin"),  // 装饰器1: 权限检查
            new TimerFilter()               // 装饰器2: 计时
        );
        if (deleteResult != null && !deleteResult.Equals(true))
        {
            Console.WriteLine($"  被取消: 返回默认值 {deleteResult}");
        }

        // --- 4. DispatchProxy 拦截器（最接近 Python 动态装饰）---
        Console.WriteLine("\n--- 4. DispatchProxy（动态代理拦截器）---");
        Console.WriteLine("  最接近 Python 装饰器的动态包装机制\n");

        var calculator = new Calculator();
        var proxy = MethodInterceptionProxy<ICalculator>.Create(calculator);
        // proxy 现在就是包装后的版本，每次调用都会被拦截

        Console.WriteLine("  调用 Add(3, 5)：");
        Console.WriteLine($"  结果: {proxy.Add(3, 5)}");

        Console.WriteLine("\n  调用 Multiply(4, 7)：");
        Console.WriteLine($"  结果: {proxy.Multiply(4, 7)}");

        // --- 总结 ---
        Console.WriteLine("\n" + new string('=', 70));
        Console.WriteLine("  总结：C# vs Python 的装饰器/AOP 对比");
        Console.WriteLine(new string('-', 70));
        Console.WriteLine("  | Python              | C#                        |");
        Console.WriteLine("  |---------------------|---------------------------|");
        Console.WriteLine("  | @decorator          | [Attribute] + 反射/框架   |");
        Console.WriteLine("  | 嵌套函数包装         | Decorator Pattern(接口)    |");
        Console.WriteLine("  | 动态替换函数         | DispatchProxy / 代理      |");
        Console.WriteLine("  | 堆叠装饰器           | Action Filter Pipeline    |");
        Console.WriteLine("  | func.__wrapped__    | 保留原始对象引用           |");
        Console.WriteLine(new string('-', 70));
        Console.WriteLine("  关键区别:");
        Console.WriteLine("  - Python: 装饰器就是函数，@ 只是语法糖");
        Console.WriteLine("  - C#: Attribute 只是元数据，需要框架配合才能拦截");
        Console.WriteLine("  - Python 更灵活（运行时包装），C# 更安全（编译时检查）");
        Console.WriteLine(new string('=', 70));
    }

    /// <summary>
    /// 用于演示 Attribute 反射读取的方法
    /// </summary>
    [Log("方法开始执行")]
    [Log("记录执行日志")]
    [Cache(DurationSeconds = 300)]
    [Authorize(Roles = "Admin")]
    public static string DemoMethod(int id)
    {
        return $"Result-{id}";
    }
}

// ============================================================================
// 7. 接口定义（供 DispatchProxy 使用）
// ============================================================================

/// <summary>
/// 计算器接口 —— DispatchProxy 必须基于接口创建
/// </summary>
public interface ICalculator
{
    int Add(int a, int b);
    int Multiply(int a, int b);
}

/// <summary>
/// 计算器实现
/// </summary>
public class Calculator : ICalculator
{
    public int Add(int a, int b) => a + b;
    public int Multiply(int a, int b) => a * b;
}

File: 第二阶段/2.5-装饰器/decorators.py
# ============================================================================
# Python 装饰器（Decorator）
# 对应文章：2.5 装饰器
# ============================================================================
#
# 【C# 程序员的 Python 修炼手册】
#
# 装饰器 = Python 版的 Attribute / 中间件 / AOP（面向切面编程）
#
# 核心思想：不修改原函数代码，给函数"套上"额外功能
#
# | C# 概念            | Python 等价             | 说明                      |
# |--------------------|-------------------------|---------------------------|
# | [Attribute]        | @decorator              | 元数据标注 / 函数包装      |
# | Action Filter      | @decorator（函数式）     | ASP.NET Core 请求拦截      |
# | 中间件 Middleware   | @decorator（栈式）       | 请求管道                   |
# | AOP 拦截器         | @decorator              | 方法前后的横切逻辑         |
# | partial class      | 类装饰器 class Decorator | 用类实现装饰器             |
#
# 关键区别：
# - C# Attribute 本质是元数据，需要反射/框架才能生效
# - Python 装饰器本质是函数包装，@ 语法糖在定义时就生效
# - Python 装饰器更灵活，但也更容易滥用
# ============================================================================

import time
import random
from functools import wraps


# ============================================================================
# 1. 最简单的装饰器 —— @timer（无括号版本）
# ============================================================================
#
# 【C# 对比】
# C# 中没有等价的无参数装饰器语法。
# C# 的 [MyAttribute] 相当于 Python 的 @my_decorator（不带括号）
#
# 什么时候用 @timer（不带括号）？
# 当装饰器只需要接收被装饰函数本身，不需要额外参数时

def timer(func):
    """
    计时装饰器：测量函数执行时间

    【为什么用 @wraps(func)？】
    没有 @wraps 时，装饰后的函数会"丢失"原函数的名字和文档：
      - __name__  变成 "wrapper" 而不是原函数名
      - __doc__   变成 wrapper 的文档而不是原函数的
      - __module__ 和其他元信息也会丢失

    这在调试和框架反射时会出大问题！
    假装你有一个被 @timer 装饰的函数叫 slow_query()，
    没有 @wraps 的话，错误日志会显示 "wrapper() failed"，
    你根本不知道是哪个函数出了问题。

    【C# 对比】
    C# 的 Attribute 不需要考虑这个问题，因为它不包装函数本身，
    只是给方法贴了个"标签"，方法还是原来的方法。
    但这也意味着 C# Attribute 无法像 Python 装饰器那样灵活地修改函数行为。
    """
    @wraps(func)  # 关键！保留原函数的 __name__, __doc__, __module__ 等
    def wrapper(*args, **kwargs):
        start = time.time()
        result = func(*args, **kwargs)
        elapsed = time.time() - start
        print(f"  [timer] {func.__name__} 执行完成，耗时: {elapsed:.4f}秒")
        return result
    return wrapper


# ============================================================================
# 2. 带参数的装饰器 —— @cache, @retry, @require_role
# ============================================================================
#
# 【C# 对比】
# 带参数的装饰器 ≈ C# Attribute 带构造函数参数
#   [Cache(DurationSeconds = 300)]
#   [Authorize(Roles = "Admin")]
#
# 但 Python 装饰器参数可以是任意表达式，比 C# Attribute 灵活得多
#
# 结构理解（三层嵌套）：
#   def decorator_with_args(arg):     # 第1层：接收装饰器参数
#       def decorator(func):          # 第2层：接收被装饰函数
#           @wraps(func)
#           def wrapper(*a, **kw):    # 第3层：实际执行的包装函数
#               ...
#           return wrapper
#       return decorator
#   return decorator

def cache(maxsize=128):
    """缓存装饰器：记忆函数结果，避免重复计算

    【C# 对比】
    等价于 [MemoryCache] 或 [Cached] 属性
    也等价于 System.Runtime.CompilerServices caching 特性
    """
    def decorator(func):
        cache_dict = {}  # 闭包捕获的缓存字典

        @wraps(func)
        def wrapper(*args):
            if args in cache_dict:
                print(f"  [cache] {func.__name__}{args} 命中缓存！")
                return cache_dict[args]
            result = func(*args)
            cache_dict[args] = result
            return result

        # 暴露缓存信息（方便调试）
        wrapper.cache_info = lambda: f"缓存大小: {len(cache_dict)}"
        wrapper.cache_clear = lambda: cache_dict.clear()
        return wrapper
    return decorator


def retry(max_attempts=3, delay=0.1):
    """重试装饰器：函数失败时自动重试

    【C# 对比】
    等价于 Polly 的 RetryPolicy
      Policy.Handle<Exception>()
            .Retry(3, (ex, i) => Console.WriteLine($"Retry {i}"))
    """
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            last_exception = None
            for attempt in range(1, max_attempts + 1):
                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    last_exception = e
                    if attempt < max_attempts:
                        print(f"  [retry] 第{attempt}次尝试失败: {e}，{delay}秒后重试...")
                        time.sleep(delay)
            raise last_exception
        return wrapper
    return decorator


def require_role(role):
    """权限检查装饰器：调用前检查用户角色

    【C# 对比】
    等价于 [Authorize(Roles = "Admin")]
    也等价于 ASP.NET Core 的 [Authorize] + Policy-based 鉴权
    """
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            user = kwargs.pop('user', None)
            if not user:
                raise PermissionError(f"[require_role] 缺少用户信息，需要 {role} 角色")
            if role not in user.get('roles', []):
                raise PermissionError(
                    f"[require_role] 用户 '{user.get('name')}' 无 {role} 角色，"
                    f"当前角色: {user.get('roles', [])}"
                )
            return func(*args, **kwargs)
        return wrapper
    return decorator


# ============================================================================
# 3. 装饰器堆叠（Stacking Decorators）
# ============================================================================
#
# 【C# 对比】
# C# 也可以堆叠 Attribute：
#   [Log("入口")]
#   [Cache(DurationSeconds = 60)]
#   [Authorize(Roles = "Admin")]
#   public void MyMethod() { ... }
#
# 执行顺序（从下往上应用，从外往内执行）：
#   @A       →    @B(func)    →  A(B(func))
#   @B
#   def func        先包装 B，再包装 A
#
# 调用时：A.wrapper → B.wrapper → func

@timer                          # 最外层，最后包装，最先执行
def heavy_computation(n):
    """一个耗时的计算函数"""
    time.sleep(0.1)
    return sum(range(n))


# ============================================================================
# 4. 类装饰器 —— Python 版的 C# Attribute / IMethodInterceptor
# ============================================================================
#
# 【C# 对比】
# 这是最接近 C# Attribute 的实现方式！
# 在 C# 中我们写：
#   [Log("开始")]
#   [Cache(Duration = 60)]
#   public void MyMethod() { ... }
#
# Python 类装饰器 vs 函数装饰器：
# - 类装饰器：可以维护状态，更接近 C# 的 class-based Attribute
# - 函数装饰器：更简洁，大多数场景够用
#
# 真正的 C# Attribute 是元数据+反射，运行时需要框架（如 ASP.NET）读取
# Python 类装饰器在 @ 语法糖处理时就直接替换函数了

class CountCalls:
    """统计函数被调用次数的类装饰器

    【C# 对比】
    类似 C# 的 ICallCounter + Autofac 注入
    但 Python 的类装饰器更直接：直接包装函数并替换
    """
    def __init__(self, func):
        wraps(func)(self)           # 把原函数的元信息复制到实例上
        self.func = func
        self.call_count = 0

    def __call__(self, *args, **kwargs):
        self.call_count += 1
        print(f"  [CountCalls] {self.func.__name__} 第{self.call_count}次调用")
        return self.func(*args, **kwargs)

    def get_stats(self):
        return f"{self.func.__name__} 已被调用 {self.call_count} 次"


# ============================================================================
# 5. 访问被包装的原始函数 —— func.__wrapped__
# ============================================================================
#
# 装饰器堆叠后，有时候我们需要绕过装饰器直接调用原函数
# Python 在 @wraps 中自动提供了 __wrapped__ 属性
#
# 【C# 对比】
# C# 没有直接等价物，但类似的设计模式：
# - Decorator Pattern 中保留原始对象的引用
# - DI 中注册原服务和装饰后的服务
# - Autofac 的 .AsSelf() / .As<T>() 注册

def debug(func):
    """调试装饰器：打印函数入参和返回值"""
    @wraps(func)
    def wrapper(*args, **kwargs):
        print(f"  [debug] 调用 {func.__name__}({args}, {kwargs})")
        result = func(*args, **kwargs)
        print(f"  [debug] 返回 {result}")
        return result
    return wrapper


# ============================================================================
# 6. 定义被装饰的函数
# ============================================================================

@timer
def slow_function():
    """模拟一个耗时操作"""
    time.sleep(0.5)
    return "完成"


@cache(maxsize=128)
def fibonacci(n):
    """斐波那契数列 —— 递归 + 缓存的经典组合"""
    if n < 2:
        return n
    return fibonacci(n - 1) + fibonacci(n - 2)


@retry(max_attempts=3, delay=0.1)
def unreliable_function():
    """模拟一个不稳定的网络调用"""
    if random.random() < 0.7:  # 70% 概率失败
        raise ConnectionError("网络连接超时")
    return "调用成功"


@require_role("Admin")
def delete_user(user_id):
    """删除用户 —— 需要 Admin 角色"""
    print(f"  已删除用户 {user_id}")
    return True


# ---- 堆叠装饰器示例 ----
@debug                          # 先打印调用信息
@timer                          # 再统计耗时（最内层）
def compute_sum(n):
    """计算 1 到 n 的和"""
    return sum(range(1, n + 1))


# ---- 类装饰器示例 ----
@CountCalls
def greet(name):
    """打招呼"""
    print(f"  Hello, {name}!")
    return f"greeted {name}"


# ============================================================================
# 主程序：演示所有装饰器效果
# ============================================================================

if __name__ == "__main__":
    print("=" * 70)
    print("  Python 装饰器（Decorator）演示")
    print("  C# 程序员的 Python 修炼手册 —— 2.5")
    print("=" * 70)

    # --- 1. @timer 无括号装饰器 ---
    print("\n--- 1. @timer（无括号装饰器）---")
    print("  说明: @timer 不带括号，直接接收函数作为参数")
    print("  C# 等价: 没有直接等价物，C# Attribute 始终需要 [] 语法\n")
    result = slow_function()
    print(f"  返回值: {result}")

    # 演示 @wraps 的效果
    print(f"\n  slow_function.__name__ = '{slow_function.__name__}'")
    print(f"  (没有 @wraps 的话这里会显示 'wrapper')")

    # --- 2. @cache 缓存装饰器 ---
    print("\n--- 2. @cache（带参数的缓存装饰器）---")
    print("  说明: @cache(maxsize=128) 带参数，需要三层嵌套")
    print("  C# 等价: [MemoryCache] 或 Polly 缓存策略\n")
    print(f"  fibonacci(10) = {fibonacci(10)}")
    print(f"  fibonacci(20) = {fibonacci(20)}")
    print(f"  {fibonacci.cache_info()}")

    # --- 3. @retry 重试装饰器 ---
    print("\n--- 3. @retry（重试装饰器）---")
    print("  说明: 失败时自动重试，C# 等价: Polly RetryPolicy\n")
    try:
        result = unreliable_function()
        print(f"  最终结果: {result}")
    except Exception as e:
        print(f"  最终失败: {e}")

    # --- 4. @require_role 权限装饰器 ---
    print("\n--- 4. @require_role（权限检查装饰器）---")
    print("  说明: C# 等价: [Authorize(Roles = \"Admin\")]\n")

    # 正确权限
    admin_user = {"name": "张三", "roles": ["Admin", "Editor"]}
    print("  尝试用 Admin 用户删除...")
    delete_user(42, user=admin_user)

    # 错误权限
    viewer_user = {"name": "李四", "roles": ["Viewer"]}
    print("\n  尝试用 Viewer 用户删除...")
    try:
        delete_user(42, user=viewer_user)
    except PermissionError as e:
        print(f"  权限不足: {e}")

    # --- 5. 装饰器堆叠 ---
    print("\n--- 5. 装饰器堆叠（Stacking Decorators）---")
    print("  @debug + @timer 同时应用在一个函数上")
    print("  执行顺序: debug.wrapper → timer.wrapper → compute_sum\n")
    result = compute_sum(1000)
    print(f"  结果: {result}")

    # --- 6. 类装饰器（C# Attribute 等价物）---
    print("\n--- 6. 类装饰器 CountCalls（C# Attribute 等价物）---")
    print("  说明: 这是最接近 C# [Attribute] 的 Python 实现方式\n")
    greet("Alice")
    greet("Bob")
    greet("Charlie")
    print(f"  统计: {greet.get_stats()}")

    # --- 7. 访问被包装的原始函数 ---
    print("\n--- 7. 访问被包装的原始函数（__wrapped__）---")
    print("  说明: 有时需要绕过装饰器，直接调用原函数\n")

    # 通过 __wrapped__ 跳过 @timer 直接调用
    print("  通过 slow_function.__wrapped__() 调用（不经过 timer）:")
    # 直接调用原函数，不会打印计时信息
    raw_start = time.time()
    raw_result = slow_function.__wrapped__()
    raw_elapsed = time.time() - raw_start
    print(f"  原始函数返回: {raw_result}，手动计时: {raw_elapsed:.4f}秒")

    print("\n" + "=" * 70)
    print("  总结：")
    print("  - @wraps 是装饰器的必备操作，保留函数元信息")
    print("  - 带参数的装饰器需要三层嵌套")
    print("  - 装饰器堆叠从下往上包装，从外往内执行")
    print("  - 类装饰器最接近 C# Attribute 的理念")
    print("  - __wrapped__ 可以绕过装饰器访问原始函数")
    print("  - Python 装饰器比 C# Attribute 更灵活，但需要自律")
    print("=" * 70)
